### PR TITLE
Remove unnecessary casts.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This is a catch-all PR that i'll be adding more commits to.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
reviewers: @balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie, @jasonmalinowski, @heejaechang

**Comment by CyrusNajmabadi** [comment]
No changes were made to the compiler layer.  In cases where we introduced an unnecessary temporary for a parameter/local, i removed the temporary and updated the code to just refer to the parameter/local instead.

**Comment by Pilchie** [comment]
Depending on what build you are running, watch out for #4531...

**Comment by CyrusNajmabadi** [comment]
@jenkins retest this please

**Comment by CyrusNajmabadi** [comment]
Thanks @Pilchie .  I think this bug did hit us in one location.  I'm going to revert that code as i'm not confident enough with it.

**Comment by CyrusNajmabadi** [comment]
Actually, it looks like it was ok to remove the casts.  But i think i may still revert things to keep the code consistent and more comprehensible.

**Comment by CyrusNajmabadi** [comment]
retest this please

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToIconFactory.cs [hunk 1]
```diff
@@ -64,7 +64,7 @@ private static Icon ConvertBitmapSourceToIcon(BitmapSource bmpSource)
 
                     // ICONDIRENTRY: Size. We'll come back to this one once we know what it actually is.
                     offsetOfSize = icoStream.Position;
-                    binaryWriter.Write((int)0);
+                    binaryWriter.Write(0);
```
Comment (DustinCampbell) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToIconFactory.cs]: Looks like the (int) cast here was there for clarity with the rest of the casts.

File: src/Workspaces/Core/Portable/SymbolId/SymbolKey.SymbolKeyComparer.cs [hunk 2]
```diff
@@ -18,12 +18,12 @@ private SymbolKeyComparer(ComparisonOptions options)
 
             public bool Equals(SymbolKey x, SymbolKey y)
             {
-                if ((object)x == (object)y)
+                if (x == (object)y)
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/SymbolId/SymbolKey.SymbolKeyComparer.cs]: Was the cast here important for reference equality or something like that? IOW, does SymbolKey overload ==?

File: src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs [hunk 3]
```diff
@@ -108,7 +108,7 @@ protected double ComputeDistance(TSequence oldSequence, int oldLength, TSequence
 
             int max = Math.Max(oldLength, newLength);
             Debug.Assert(lcsLength <= max);
-            return 1.0 - (double)lcsLength / (double)max;
+            return 1.0 - lcsLength / (double)max;
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs]: Without engaging my brain, this one scares me a bit.

**Comment by DustinCampbell** [comment]
I scanned through, and didn't have any questions on 99% of them. Out of curiosity, while performing this exercise, did you find any false positives for Remove Redundant Cast?

File: src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToIconFactory.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/NavigateTo/NavigateToIconFactory.cs]: fair point.

File: src/Workspaces/Core/Portable/SymbolId/SymbolKey.SymbolKeyComparer.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/SymbolId/SymbolKey.SymbolKeyComparer.cs]: nope.  The cast is redundant.  only one side needs to be object.  honestly though, i'd rather the code call ReferenceEquals

File: src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs]: i'll revert it.

**Comment by CyrusNajmabadi** [comment]
Nope.  No false positive.

**Pull request merged**